Add docs for Slices API (#36612)
gatsbyjs/gatsby: docs/docs/reference/built-in-components/gatsby-slice.md

    context: {
      jokeOfTheDay: `What's blue and not heavy? Light blue.`,
    }
  })
}
```

| Argument               | Type     | Description                                                                |
| ---------------------- | -------- | -------------------------------------------------------------------------- |
| `id` (Required)        | `string` | A unique identifier for this specific Slice. See also: [Aliases](#aliases) |
| `component` (Required) | `string` | The path to the component being used as a Slice.                           |
| `context`              | `object` | An object passed to the `component` as `sliceContext`.                     |

## `Slice` Component

The Slice component requires an `alias` prop. Any props additional props will be passed to underlying component.

| Prop               | Type     | Description                                                                                 |
| ------------------ | -------- | ------------------------------------------------------------------------------------------- |
| `alias` (Required) | `string` | The Slice created in `gatsby-node` to replace this component. See also: [Aliases](#aliases) |

> ### Restrictions on using `Slice`

Author: Should also add a restriction for nested slices

Reviewer: Move the restrictions to the end after explaining the API so that the descriptions make sense to people

---

> title: Gatsby Slice API

Reviewer: Is it "Gatsby Slice" API or "Gatsby Slice**s**" API?

Author: I was going off the naming pattern of other components. I know our RFC calls it "Slices", but I think matching the others (i.e. Gatsby Script API) is the best path forward.

---

> Support for the Gatsby Slice API was added in `gatsby@5.0.0`.

> To further the improvements seen by [Incremental Builds](https://www.gatsbyjs.com/blog/2020-04-22-announcing-incremental-builds/), Gatsby includes a built-in `<Slice>` component to divide pages into incrementally built "slices".

Reviewer 2: ```suggestion
To further the improvements seen by [Incremental Builds](https://www.gatsbyjs.com/blog/2020-04-22-announcing-incremental-builds/), Gatsby includes a built-in `<Slice>` component to divide parts of pages into incrementally built "slices".
```

> To further the improvements seen by [Incremental Builds](https://www.gatsbyjs.com/blog/2020-04-22-announcing-incremental-builds/), Gatsby includes a built-in `<Slice>` component to divide pages into incrementally built "slices".
> 
> By using a new `<Slice>` React component in combination with the [`createSlice`](/docs/reference/config-files/actions/#createNodeField) API for common UI features, Gatsby will be able to build and deploy individual pieces of your site that had content changes, not just entire pages.

Reviewer: I think we can explain the concept better without resorting to quotes or re-using the word itself. It won't click for everyone what `"slices"` means.

So we'll need to describe in simpler words what that "slices" would mean. I don't have a perfect sentence yet but here would be a start:

```suggestion
To further the improvements seen by [Incremental Builds](https://www.gatsbyjs.com/blog/2020-04-22-announcing-incremental-builds/), Gatsby includes a built-in `<Slice>` component that allows you to split pages into individual parts.

By using the `<Slice>` React component in combination with the [`createSlice`](/docs/reference/config-files/actions/#createSlice) API for common UI features, Gatsby will be able to build and deploy individual pieces of your site that had content changes, not just entire pages.
```

> ## `<Slice>` enables faster builds in Gatsby Cloud

Reviewer: ```suggestion
## Faster builds in Gatsby Cloud
```

With the introduction of [Incremental Builds](https://www.gatsbyjs.com/blog/2020-04-22-announcing-incremental-builds/), Gatsby Cloud has been able to reduce build times signficantly by only building the pages that changed. The `<Slice>` component helps reduce those builds times further.

> Common components that are shared across the majority of pages on your site might include a navigation bar, footer, or contact form. In today's frameworks, when you re-order the navigation bar items, the entire site needs to be rebuilt. However, if the navigation bar was created as a Gatsby Slice, the new navigation items only need to built once and all pages will pull the new navigation bar when it's needed.

Reviewer 2: ```suggestion
Common components that are shared across the majority of pages on your site might include a navigation bar, footer, or contact form. In today's frameworks, when you re-order the navigation bar items, the entire site needs to be rebuilt. However, if the navigation bar was created as a Gatsby Slice, the new navigation items only need to be built once and all pages will pull the new navigation bar when it's needed.
```

> ## How to use Gatsby's `<Slice>`

Reviewer: ```suggestion
## Using Gatsby Slice in your site
```

For this example, let's use the same scenario as described above - a large site that has a shared navigation bar.

> ```typescript:title=src/layouts/default-layout.tsx

Reviewer: Our code examples are all JS, so this needs to be changed here

import { NavigationBar, Footer } from "../components"

export const DefaultLayout = ({ children, navigationBarClassName }: DefaultLayoutProps) => {

>   /// ...

Reviewer: Adding those `/// ...` is adding extra noise, we should remove them in all code examples

  return (

>     <div className={styles.defaultLayout} />

Reviewer 2: ```suggestion
    <div>
```

Extra styles not relevant to the example, remove self closing tag

      <NavigationBar className={navigationBarClassName} />
      {content}
      <Footer />
    </div>
  )
}
```

> ### Creating the Slice in `gatsby-node`

Reviewer: This whole "How to use Gatsby's `<Slice>`" kinda reads like a step-by-step how-to guide (https://www.gatsbyjs.com/contributing/docs-contributions/how-to-write-a-how-to-guide/), not a reference guide (https://www.gatsbyjs.com/contributing/docs-contributions/how-to-write-a-reference-guide/)

The "How to use" should be condensed down into two code blocks, first one showing the `createSlice` API, second showing how to use the slice in the page.

Creating a Gatsby Slice is simple using the [`createSlice`](/docs/reference/config-files/actions/#createNodeField) action from the [`createPages`](/docs/reference/config-files/gatsby-node/#createPages) API in your `gatsby-node`.

```typescript:title=gatsby-node.ts
exports.createPages = async ({ actions }) => {
  actions.createSlice({
    id: `navigation-bar`,
    component: require.resolve(`./src/components/navigation-bar.js`),
  })

  // ...
}
```

Now that we have a `navigation-bar` Slice created, let's go use it!

### Using created Slices

Now we need to convert the `DefaultLayout` component to actually use the new Slice we created.

```diff
-import { NavigationBar, Footer } from "../components"
+import { Footer } from "../components"
+import { Slice } from "gatsby"

export const DefaultLayout = ({ children, navigationBarClassName }: DefaultLayoutProps) => {
  /// ...

  return (
    <div className={styles.defaultLayout} />
-     <NavigationBar className={navigationBarClassName} />

> +     <Slice alias="navigation-bar" className={navigationBarClassName} />

Reviewer: Are there any things that are worth mentioning about how to use the Slice API in pages? Would it warrant its own section?

      {content}
      <Footer />
    </div>
  )
}
```

> That's it! After a successful `yarn gatsby build`, you should see a list of `Slices` that were built for your site.

Reviewer 2: ```suggestion
That's it! After a successful `gatsby build`, you should see a list of `Slices` that were built for your site.
```

Reviewer 2: > you should see a list of `Slices` that were built for your site

Does this mean a list of created slices are shown in the reporter output?

Author: > Does this mean a list of created slices are shown in the reporter output?

Yep! We show a list just under the "Pages" output of builds.

> ### Using Context

Reviewer: Using context and using aliases should be grouped under the h2 of `createSlice`

Similar to the context that can be passed to pages in [`createPages`](/docs/reference/config-files/gatsby-node/#createPages), [`createSlice`](/docs/reference/config-files/actions/#createNodeField) can also pass context to individual slices.

```typescript:title=gatsby-node.ts
exports.createPages = async ({ actions }) => {
  actions.createSlice({
    id: `navigation-bar`,
    context: {
      jokeOfTheDay: `What's blue and not heavy? Light blue.`,
    }
    component: require.resolve(`./src/components/navigation-bar.js`),
  })

  // ...
}
```

The data passed to `context` here will be handed down to the `NavigationBar` component with the key `sliceContext`.

```typescript:title=src/components/navigation-bar.tsx
// highlight-next-line
export const NavigationBar = ({ className, sliceContext }: NavigationBarProps) => {
  // ...

  return (
    <div className={className}>
      <Link to="/">Home</Link>
      // highlight-next-line
      <Link to="/jokes">{sliceContext.jokeOfTheDay}</Link>
    </div>
  )
}
```

### Using Aliases

There will be times where a single Slice will either need to be handed different context or swapped entirely depending on which page it's being rendered on.

> This is where we can utilize the `alias` prop that is given to the `<Slice>` component. When we converted the `<NavigationBar>` component to a slice, we created the slice with an `id` of `navigation-bar`, but passed that value to the `alias` prop of `<Slice>`. Why is there a difference in key names?

Reviewer: Haven't checked all text but instead of "we", use "you"
https://www.gatsbyjs.com/contributing/gatsby-style-guide/#use-you-as-the-pronoun

> An `alias` is not a 1-to-1 mapping of string-to-slice. When we create a page using [`createPages`](/docs/reference/config-files/gatsby-node/#createPages), we can actually pass a key-value map of alias-to-id to tell Gatsby which Slice to use throughout the page.

Reviewer: These two sentences have too much jargon for my taste (https://www.gatsbyjs.com/contributing/gatsby-style-guide/#define-jargon) and I had difficulties understanding it

| `id` (Required)        | `string` | A unique identifier for this specific Slice. See also: [Aliases](#aliases) |
| `component` (Required) | `string` | The path to the component being used as a Slice.                           |
| `context`              | `object` | An object passed to the `component` as `sliceContext`.                     |

## `Slice` component

The Slice component requires an `alias` prop. Any props additional props will be passed to underlying component.

| Prop               | Type     | Description                                                                                 |
| ------------------ | -------- | ------------------------------------------------------------------------------------------- |
| `alias` (Required) | `string` | The Slice created in `gatsby-node` to replace this component. See also: [Aliases](#aliases) |

### Restrictions on using `Slice`

#### Nested Slices

Gatsby does not support nested Slices. This means if you have a high level `<Layout>` component as a slice, other Slices cannot exist within that `<Layout>` component anywhere in the tree.

> #### Props

Reviewer: Shouldn't this be an h3?

Author: This is talking about restrictions of props (under the "Restrictions on using `Slice`" title).

Reviewer: ahh 👍
